Add a blogpost "State of particles and future updates" (#734)
godotengine/godot-website: collections/_article/progress-report-state-of-particles.md

---
title: "State of particles and future updates"
excerpt: "An update on the current state of the particle system in Godot, and some hints about what the future holds!"
categories: ["progress-report"]
author: Ilaria Cislaghi
image: /storage/blog/covers/progress-report-state-of-particles.webp

> date: 2023-10-23 15:00:00

Author: ```suggestion
date: 2023-10-23 13:00:00
```

Multiple configuration options were added or changed to expand control over damping, emission, and velocity.

Damping changes include:

- `damping_as_friction`: flag that changes the behavior of damping from a constant
deceleration to a deceleration based on speed.

Emission changes include:

- `emission_shape_offset`: an offset for the emission shape, in the node's local space.
- `emission_shape_scale`: the scale of the emission shape, in the node's local space.
- `inherit_velocity_ratio`: a percentage of the emitter's velocity to inherit on spawn.

Velocity changes include:

- `velocity_pivot`: a pivot point used for calculating radial and orbital velocity.
- `directional_velocity`: velocity along an axis; requires an XYZ curve.
- `orbit_velocity`: orbit velocity can now be used without toggling `disable_z`; requires an XYZ curve; axes are in the node's local space.
- `radial_velocity`: velocity away or towards the velocity pivot.
- `velocity_limit`: a curve defining a hard limit for the particles' velocity following their lifetime.
- `scale_over_velocity`: a curve that allows to scale particles on a given axis.

> On top of that an optional separation between alpha over lifetime and emission over lifetime was added. Note that emission over lifetime just multiplies the `COLOR` value of the particle by a given value. This needs to be properly read in the display shader.

Reviewer: ```suggestion
On top of that, an optional separation between alpha over lifetime and emission over lifetime was added. Note that emission over lifetime just multiplies the `COLOR` value of the particle by a given value. This needs to be properly read in the display shader.
```

You can also check out [GH-79527](https://github.com/godotengine/godot/pull/79527), which covers the bulk of these changes.

I can't wait to see all that you'll make with it! In the meantime, enjoy these VFX made by the testers of the PR.

<video autoplay loop muted playsinline title="A compilation of various effects created by community members during testing">
  <source src="/storage/blog/2023-oct-state-of-particles/particle-fiesta-by-community.mp4" type="video/mp4">
</video>

> _Various particle effects created by community members to help test the rework. Credits go to [Tibo](https://gotibo.fr/), [Calinou](https://github.com/Calinou), and myself, [QbieShay](https://twitter.com/QbieShay)_

Reviewer 2: ```suggestion
_Various particle effects created by community members to help test the rework. Credits go to [Thibaud](https://gotibo.fr/), [Calinou](https://github.com/Calinou), and myself, [QbieShay](https://social.sparkles.cafe/@qbie)_
```

Accordingly, we now see the `CPUParticles` as a lower-end, simpler alternative to GPU particles, instead of a CPU-based, 1-to-1 fallback. It's likely that no more features will be added to `CPUParticles`. We recommend migrating your effects to `GPUParticles`. To help you with the migration, [a converter from CPU particles to a GPU equivalent](https://github.com/godotengine/godot/pull/80779) was added.

From Godot 4.2 onwards, new features for `CPUParticles` will not be accepted unless they have a counterpart in `GPUParticles`. However, contributions to bring CPU particles to feature parity are welcome.

## The future development

> For now, development around particles will mostly focus on fixing up recent changes. At the top of my personal priority list, as the area maintainer, are more spawn options (circles, lines, helixes) and more control over the ordering of particle spawning. This should make it possible to spawn particles orderly when necessary, possibly tweaking randomness to add optional variety or stricter order to the way particles are spawned.

Reviewer 2: ```suggestion
For now, development around particles will continue focusing on fixing any bug that may have gotten past the review step of this feature.
While a lot of people tested the PR, there can be still some less common particle features that still have issues.
Thanks everyone that helped with the test and the development! It's been a huge undertaking and it wouldn't have been possible without every single one of you <3
```

> After these features are implemented, I'd consider the built-in particle material good enough for beginner to medium use. For advanced use, the particle process material supports visual shaders, and in the future I'd love to see more visual shader nodes added.

Reviewer 2: ```suggestion
I am taking time to recollect and assess the current needs of the vfx pipeline in Godot.
My goal, first and foremost, is to enable an artist-friendly workflow, which means reducing the technicalities and the "gotchas" that Godot currently has on that front.
In no specific order, there's a couple of things on my mind such as the visual shader support for process materials, extra visual shader nodes, more spawn options for particles, etc.
You can check the [VFX and Techart wishlist](https://github.com/orgs/godotengine/projects/54) on Github for more information on that.
```

---
title: "State of particles and future updates"
excerpt: "An update on the current state of the particle system in Godot, and some hints about what the future holds!"
categories: ["progress-report"]
author: Ilaria Cislaghi
image: /storage/blog/covers/progress-report-state-of-particles.webp
date: 2023-10-17 14:00:00
---

_TL;DR: In 4.2 GPUParticles were refactored and lots of new features were added. Due to high maintenance cost and bug surface, CPUParticles will keep the current feature set._

-----

Godot 4 offers a great opportunity to make some long-requested and much needed changes in many areas, and one of those areas is particle nodes and resources. Before 4.0 some settings used to be a bit confusing and clunky to work with. Now a lot of things have changed: parameters are set with a range using min and max boundaries instead of value and random deviation; there's a particle collision system; there's a turbulence system, there are sub-emitters, and much more.

> All around, particles already received a lot of love during the development of Godot 4. However, a number of key features were still missing as of 4.1. Below I go into detail about additions and improvements made to the particle system in the last few months.

Reviewer 3: ```suggestion
All around, particles already received a lot of love during the development of Godot 4. However, a number of key features were still missing as of 4.1. Below I go into detail about additions and improvements made to the particle system in the last few months, which you can test already in the [beta release of Godot 4.2](/article/dev-snapshot-godot-4-2-beta-1/).
```

<video autoplay loop muted playsinline title="A countdown with particle effects">
  <source src="/storage/blog/state-of-particles/final-countdown.mp4" type="video/mp4">
</video>

## Artistic control

> Godot's existing particle system made an assumption that people would use it for more physically accurate simulations, such which are suited for more realistic games. But in practice a vast majority of projects in the Godot ecosystem choose a stylized art direction.

Reviewer 4: ```suggestion
Godot's existing particle system made an assumption that people would use it for more physically accurate simulations, which are suited for more realistic games. But in practice a vast majority of projects in the Godot ecosystem choose a stylized art direction.
```

> This creates a big pain point when using current particles. In order to obtain an exaggerated motion and snappy timings in visual effects, acceleration-based approach just doesn't cut it. It ends up really clunky, hard to use, and sometimes achieving certain effects is just not possible.

Reviewer 3: Too many "just" (which is btw a dis-recommended word like "simply", but here I think it's fine).
```suggestion
This creates a big pain point when using current particles. In order to obtain an exaggerated motion and snappy timings in visual effects, an acceleration-based approach just doesn't cut it. It ends up really clunky, hard to use, and sometimes achieving certain effects is not possible.
```

> To address these issues, and more, the `GPUParticles` system has been refactored and expanded upon. Particles can now support not just realistic types of effects, but also more stylized ones with more exaggerated motion. Additionally, a couple of long-requested quality of life features was added.

Reviewer 3: ```suggestion
To address these issues, and more, the `GPUParticles` system has been refactored and expanded upon. Particles can now support not only realistic types of effects, but also more stylized ones with more exaggerated motion. Additionally, a couple of long-requested quality of life features was added.
```

> To address these issues, and more, the `GPUParticles` system has been refactored and expanded upon. Particles can now support not just realistic types of effects, but also more stylized ones with more exaggerated motion. Additionally, a couple of long-requested quality of life features was added.

Reviewer 5: ```suggestion
To address these issues, and more, the `GPUParticles` system has been refactored and expanded upon. Particles can now support not just realistic types of effects, but also more stylized ones with more exaggerated motion. Additionally, a couple of long-requested quality of life features were added.
```

## The new `GPUParticles`

**NOTE:** These changes should be 100% backwards compatible. If particles look different after updating, please [file a bug report](https://github.com/godotengine/godot/issues/new/choose).

> _[Pioneering of this work](https://github.com/godotengine/godot/pull/78851) was kindly sponsored by [PahdoLabs](https://www.pahdolabs.com/)._

Reviewer 3: It might be a bit confusing that this links to an unmerged PR. Maybe it should link both 78851 as the one pioneering that work, and then 79527? E.g.:
```suggestion
_[Pioneering of this work](https://github.com/godotengine/godot/pull/78851) was kindly sponsored by [Pahdo Labs](https://www.pahdolabs.com/)._ After further iteration, this was eventually merged in [GH-79527](https://github.com/godotengine/godot/pull/79527).
```
(Also space in "Pahdo Labs", at least judging by their website.)

### New parameters for `GPUParticles` nodes

Two new parameters were added to `GPUParticles2D`/`GPUParticles3D` nodes:

- `amount_ratio`: from 0 to 1, a percentage of particles compared to the amount that will be emitted.
- `interp_to_end`: from 0 to 1, allows to interpolate all particles from the same node simultaneously towards the end of their lifetime.

### Changes in `ParticleProcessMaterial`

The main change in the particle material is the big-big refactor of the internal code. The code was split into smaller, reusable functions, which are

> called both from `start()` and `process()`. This ensures better consistency and more maintainability.

Reviewer 6: ```suggestion
called both from `start()` and `process()`. This ensures better consistency and maintainability.
```

> Multiple configuration options were added or changed to expand your control over damping, emission, and velocity.

Reviewer 6: ```suggestion
Multiple configuration options were added or changed to expand control over damping, emission, and velocity.
```

Author: I mean, we're talking to the reader 🙃 But fine

Damping changes include:

- `damping_as_friction`: flag that changes the behavior of damping from a constant
deceleration to a deceleration based on speed.

Emission changes include:

- `emission_shape_offset`: an offset for the emission shape, in the node's local space.

> - `emission_shape_scale`: a scale of the emission shape, in the node's local space.

Reviewer 3: ```suggestion
- `emission_shape_scale`: the scale of the emission shape, in the node's local space.
```

- `inherit_velocity_ratio`: a percentage of the emitter's velocity to inherit on spawn.

Velocity changes include:

- `velocity_pivot`: a pivot point used for calculating radial and orbital velocity.
- `directional_velocity`: velocity along an axis; requires an XYZ curve.
- `orbit_velocity`: orbit velocity can now be used without toggling `disable_z`; requires an XYZ curve; axes are in the node's local space.
- `radial_velocity`: velocity away or towards the velocity pivot.
- `velocity_limit`: a curve defining a hard limit for the particles' velocity following their lifetime.
- `scale_over_velocity`: a curve that allows to scale particles on a given axis.

> On top of that an optional separation between alpha over life and emission over life was added. Note that emission over lifetime just multiplies the `COLOR` value of the particle by a given value. This needs to be properly read in the display shader.

Reviewer 5: ```suggestion
On top of that an optional separation between alpha over lifetime and emission over lifetime was added. Note that emission over lifetime just multiplies the `COLOR` value of the particle by a given value. This needs to be properly read in the display shader.
```

You can also check out [GH-79527](https://github.com/godotengine/godot/pull/79527), which covers the bulk of these changes.

> I can't wait to see all that you'll make with it! In the meantime, enjoy these VFX made by the testers of the PR.

Author: So @QbieShay wants to add a video by Rensei here. Still waiting on a link or a file with it.

Reviewer 2: I like the countdown at the top, here we can add the rest of the videos from tibo, myself, calinou and tuppy

## And what about `CPUParticles`?

> CPU particles have been maintained alongside GPU particles for a long time, but it's becoming [harder and harder to keep feature parity](https://github.com/godotengine/godot-proposals/issues/7344), especially after big changes such as the one described above. CPU particles were added at the times of GLES2 renderer in Godot 3, which did not have the capability of supporting `GPUParticles`. Godot 4 has Vulkan and GLES3 renderers, both of which support GPU particles. This makes it hard to justify maintaining 4 separate nodes in 2 different languages.

Reviewer 3: ```suggestion
CPU particles have been maintained alongside GPU particles for a long time, but it's becoming [harder and harder to keep feature parity](https://github.com/godotengine/godot-proposals/issues/7344), especially after big changes such as the one described above. CPU particles were added at the time of the GLES2 renderer in Godot 3, which did not have the capability to support GPU particles. Godot 4 has Vulkan and GLES3 renderers, both of which support GPU particles. This makes it hard to justify maintaining 4 separate nodes in 2 different languages (GLSL and C++).
```

> CPU particles have been maintained alongside GPU particles for a long time, but it's becoming [harder and harder to keep feature parity](https://github.com/godotengine/godot-proposals/issues/7344), especially after big changes such as the one described above. CPU particles were added at the times of GLES2 renderer in Godot 3, which did not have the capability of supporting `GPUParticles`. Godot 4 has Vulkan and GLES3 renderers, both of which support GPU particles. This makes it hard to justify maintaining 4 separate nodes in 2 different languages.

Reviewer 5: ```suggestion
CPU particles have been maintained alongside GPU particles for a long time, but it's becoming [harder and harder to keep feature parity](https://github.com/godotengine/godot-proposals/issues/7344), especially after big changes such as the ones described above. CPU particles were added at the time of the GLES2 renderer in Godot 3, which did not have the capability of supporting `GPUParticles`. Godot 4 has Vulkan and GLES3 renderers, both of which support GPU particles. This makes it hard to justify maintaining four separate nodes in two different languages.
```

> Unfortunately, unless there are people willing to maintain `CPUParticles`, they will begin to diverge further from `GPUParticles` in future releases. It's likely that no more features will be added to them. We recommend migrating your effects to `GPUParticles`. To help you with the migration, [a converter from CPU particles to a GPU equivalent](https://github.com/godotengine/godot/pull/80779) was added.

Reviewer 5: ```suggestion
Unfortunately, unless there are people willing to maintain `CPUParticles`, they will begin to diverge further from `GPUParticles` in future releases. It's likely that no more features will be added to `CPUParticles`. We recommend migrating your effects to `GPUParticles`. To help you with the migration, [a converter from CPU particles to a GPU equivalent](https://github.com/godotengine/godot/pull/80779) was added.
```

> Unfortunately, unless there are people willing to maintain `CPUParticles`, they will begin to diverge further from `GPUParticles` in future releases. It's likely that no more features will be added to `CPUParticles`. We recommend migrating your effects to `GPUParticles`. To help you with the migration, [a converter from CPU particles to a GPU equivalent](https://github.com/godotengine/godot/pull/80779) was added.

Reviewer 4: ```suggestion
Accordingly, we now see the `CPUParticles` as a lower-end, simple alternative to GPUParticles, instead of a CPU-based, 1-to-1 fallback. It's likely that no more features will be added to `CPUParticles`. We recommend migrating your effects to `GPUParticles`. To help you with the migration, [a converter from CPU particles to a GPU equivalent](https://github.com/godotengine/godot/pull/80779) was added.
```

This sounds like it is blaming the reader for not doing more. I don't think there is any need for this ultimatum anyway

Reviewer 2: The way you changed it though sounds like we won't accept people that backport features to cpuparticles, i think. I don't want to blame the readers of course but I would like to also not close the door if someone wants to do that work

Author: I agree, good suggestion! I couldn't figure out what to do with this sentence myself. (Edit: This refers to Clay's comment; Qbie's was added after mine, but dates to 4 days ago)

Author: I adjusted the new message to mention welcoming PRs for parity; and also to make it sound more like Qbie :)
